[BUG] jsonMapper compilation failed if model has properties with name equal to reserved keyword (example: enum)

### Describe the bug

Following v7.36 release, the new jsonMapper compilation failed if property name of the model are equal to reserved keyword

```
[2023-09-27T16:06:49.479] [ERROR] [TSED] - {
  method: 'GET',
  url: '/test/failed',
  route: '/test/failed',
  headers: {
    host: 'localhost:8083',
    'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/117.0',
    accept: 'application/json',
    'accept-language': 'fr,fr-FR;q=0.8,en-US;q=0.5,en;q=0.3',
    'accept-encoding': 'gzip, deflate, br',
    dnt: '1',
    connection: 'keep-alive',
    cookie: '_pk_id.1.1fff=b1f543f220c17ea0.1689800215.',
    'sec-fetch-dest': 'empty',
    'sec-fetch-mode': 'cors',
    'sec-fetch-site': 'same-origin'
  },
  body: {},
  query: {},
  params: {},
  reqId: '35313dbaf326466488c91321b5709fc1',
  time: 2023-09-27T14:06:49.479Z,
  duration: 42,
  event: 'request.end',
  status: 500,
  status_code: '500',
  state: 'KO',
  error_name: 'Error',
  error_message: 'Fail to compile mapper for TestModel. See the error above: Unexpected strict mode reserved word.
' +
    '(input,options) => {
' +
    '  if (isNil(input)) {
' +
    '    return input;
' +
    '  }
' +
    '  const obj = {};
' +
    '  options = {...options, self: input};
' +
    '  if (input && isCollection(input)) {
' +
    "    return execMapper(nameOf(classOf(input)), input, {...options, id: 'TestModel:eebzbch:default'});
" +
    '  }
' +
    '  if (hasJsonMethod(input)) {
' +
    "    return mapJSON(input, {...options, id: 'TestModel:eebzbch:default'});
" +
    '  }
' +
    '  // Map type TestModel:eebzbch:default 
' +
    '  let type = input.type;
' +
    '  if (type !== undefined) {
' +
    "    type = execMapper('String', type, options);
" +
    '    obj.type = type;
' +
    '  }
' +
    '  // Map enum TestModel:eebzbch:default 
' +
    '  let enum = input.enum;
' +
    '  if (enum !== undefined) {
' +
    "    enum = execMapper('Array', enum, { ...options, id: 'Object' });
" +
    '    obj.enum = enum;
' +
    '  }
' +
    '  // Map default TestModel:eebzbch:default 
' +
    '  let default = input.default;
' +
    '  if (default !== undefined) {
' +
    "    default = execMapper('String', default, options);
" +
    '    obj.default = default;
' +
    '  }
' +
    '  // Map name TestModel:eebzbch:default 
' +
    '  let name = input.name;
' +
    '  if (name !== undefined) {
' +
    "    name = execMapper('String', name, options);
" +
    '    obj.name = name;
' +
    '  }
' +
    '  // Map description TestModel:eebzbch:default 
' +
    '  let description = input.description;
' +
    '  if (description !== undefined) {
' +
    "    description = execMapper('String', description, options);
" +
    '    obj.description = description;
' +
    '  }
' +
    '  return obj;
' +
    '}',
  error_stack: 'Error: Fail to compile mapper for TestModel. See the error above: Unexpected strict mode reserved word.
' +
    '(input,options) => {
' +
    '  if (isNil(input)) {
' +
    '    return input;
' +
    '  }
' +
    '  const obj = {};
' +
    '  options = {...options, self: input};
' +
    '  if (input && isCollection(input)) {
' +
    "    return execMapper(nameOf(classOf(input)), input, {...options, id: 'TestModel:eebzbch:default'});
" +
    '  }
' +
    '  if (hasJsonMethod(input)) {
' +
    "    return mapJSON(input, {...options, id: 'TestModel:eebzbch:default'});
" +
    '  }
' +
    '  // Map type TestModel:eebzbch:default 
' +
    '  let type = input.type;
' +
    '  if (type !== undefined) {
' +
    "    type = execMapper('String', type, options);
" +
    '    obj.type = type;
' +
    '  }
' +
    '  // Map enum TestModel:eebzbch:default 
' +
    '  let enum = input.enum;
' +
    '  if (enum !== undefined) {
' +
    "    enum = execMapper('Array', enum, { ...options, id: 'Object' });
" +
    '    obj.enum = enum;
' +
    '  }
' +
    '  // Map default TestModel:eebzbch:default 
' +
    '  let default = input.default;
' +
    '  if (default !== undefined) {
' +
    "    default = execMapper('String', default, options);
" +
    '    obj.default = default;
' +
    '  }
' +
    '  // Map name TestModel:eebzbch:default 
' +
    '  let name = input.name;
' +
    '  if (name !== undefined) {
' +
    "    name = execMapper('String', name, options);
" +
    '    obj.name = name;
' +
    '  }
' +
    '  // Map description TestModel:eebzbch:default 
' +
    '  let description = input.description;
' +
    '  if (description !== undefined) {
' +
    "    description = execMapper('String', description, options);
" +
    '    obj.description = description;
' +
    '  }
' +
    '  return obj;
' +
    '}
' +
    '    at JsonSerializer.compile (/Users/workspace/test7.36/node_modules/@tsed/json-mapper/src/domain/JsonMapperCompiler.ts:168:15)
' +
    '    at JsonSerializer.map (/Users/workspace/test7.36/node_modules/@tsed/json-mapper/src/domain/JsonSerializer.ts:60:25)
' +
    '    at serialize (/Users/workspace/test7.36/node_modules/@tsed/json-mapper/src/utils/serialize.ts:11:21)
' +
    '    at PlatformResponseFilter.serialize (/Users/workspace/test7.36/node_modules/@tsed/platform-response-filter/src/services/PlatformResponseFilter.ts:92:25)
' +
    '    at PlatformHandler.flush (/Users/workspace/test7.36/node_modules/@tsed/common/src/services/PlatformHandler.ts:98:44)
' +
    '    at /Users/workspace/test7.36/node_modules/@tsed/common/src/services/PlatformHandler.ts:184:51
' +
    '    at catchAsyncError (/Users/workspace/test7.36/node_modules/@tsed/core/src/utils/catchError.ts:11:11)
' +
    '    at PlatformHandler.onFinish (/Users/workspace/test7.36/node_modules/@tsed/common/src/services/PlatformHandler.ts:184:39)
' +
    '    at /Users/workspace/test7.36/node_modules/@tsed/platform-express/src/components/PlatformExpress.ts:152:13
' +
    '    at runInContext (/Users/workspace/test7.36/node_modules/@tsed/di/src/node/utils/asyncHookContext.ts:24:12)'
}
```

### To Reproduce

1. Clone https://github.com/Bugslive/tsed-7.36.git
2. run 'npm run build'
3. run 'npm start'
4. Perform GET http://localhost:8083/test => 200 OK
5. Perform GET http://localhost:8083/test/failed => 500 KO

### Expected behavior

Ts.ED jsonMapper should retrieve v7.35 behaviour and accept to serialize/deserialize json with this kind of property name. 

### Code snippets

```
[2023-09-27T16:06:49.479] [ERROR] [TSED] - {
  method: 'GET',
  url: '/test/failed',
  route: '/test/failed',
  headers: {
    host: 'localhost:8083',
    'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/117.0',
    accept: 'application/json',
    'accept-language': 'fr,fr-FR;q=0.8,en-US;q=0.5,en;q=0.3',
    'accept-encoding': 'gzip, deflate, br',
    dnt: '1',
    connection: 'keep-alive',
    cookie: '_pk_id.1.1fff=b1f543f220c17ea0.1689800215.',
    'sec-fetch-dest': 'empty',
    'sec-fetch-mode': 'cors',
    'sec-fetch-site': 'same-origin'
  },
  body: {},
  query: {},
  params: {},
  reqId: '35313dbaf326466488c91321b5709fc1',
  time: 2023-09-27T14:06:49.479Z,
  duration: 42,
  event: 'request.end',
  status: 500,
  status_code: '500',
  state: 'KO',
  error_name: 'Error',
  error_message: 'Fail to compile mapper for TestModel. See the error above: Unexpected strict mode reserved word.
' +
    '(input,options) => {
' +
    '  if (isNil(input)) {
' +
    '    return input;
' +
    '  }
' +
    '  const obj = {};
' +
    '  options = {...options, self: input};
' +
    '  if (input && isCollection(input)) {
' +
    "    return execMapper(nameOf(classOf(input)), input, {...options, id: 'TestModel:eebzbch:default'});
" +
    '  }
' +
    '  if (hasJsonMethod(input)) {
' +
    "    return mapJSON(input, {...options, id: 'TestModel:eebzbch:default'});
" +
    '  }
' +
    '  // Map type TestModel:eebzbch:default 
' +
    '  let type = input.type;
' +
    '  if (type !== undefined) {
' +
    "    type = execMapper('String', type, options);
" +
    '    obj.type = type;
' +
    '  }
' +
    '  // Map enum TestModel:eebzbch:default 
' +
    '  let enum = input.enum;
' +
    '  if (enum !== undefined) {
' +
    "    enum = execMapper('Array', enum, { ...options, id: 'Object' });
" +
    '    obj.enum = enum;
' +
    '  }
' +
    '  // Map default TestModel:eebzbch:default 
' +
    '  let default = input.default;
' +
    '  if (default !== undefined) {
' +
    "    default = execMapper('String', default, options);
" +
    '    obj.default = default;
' +
    '  }
' +
    '  // Map name TestModel:eebzbch:default 
' +
    '  let name = input.name;
' +
    '  if (name !== undefined) {
' +
    "    name = execMapper('String', name, options);
" +
    '    obj.name = name;
' +
    '  }
' +
    '  // Map description TestModel:eebzbch:default 
' +
    '  let description = input.description;
' +
    '  if (description !== undefined) {
' +
    "    description = execMapper('String', description, options);
" +
    '    obj.description = description;
' +
    '  }
' +
    '  return obj;
' +
    '}',
  error_stack: 'Error: Fail to compile mapper for TestModel. See the error above: Unexpected strict mode reserved word.
' +
    '(input,options) => {
' +
    '  if (isNil(input)) {
' +
    '    return input;
' +
    '  }
' +
    '  const obj = {};
' +
    '  options = {...options, self: input};
' +
    '  if (input && isCollection(input)) {
' +
    "    return execMapper(nameOf(classOf(input)), input, {...options, id: 'TestModel:eebzbch:default'});
" +
    '  }
' +
    '  if (hasJsonMethod(input)) {
' +
    "    return mapJSON(input, {...options, id: 'TestModel:eebzbch:default'});
" +
    '  }
' +
    '  // Map type TestModel:eebzbch:default 
' +
    '  let type = input.type;
' +
    '  if (type !== undefined) {
' +
    "    type = execMapper('String', type, options);
" +
    '    obj.type = type;
' +
    '  }
' +
    '  // Map enum TestModel:eebzbch:default 
' +
    '  let enum = input.enum;
' +
    '  if (enum !== undefined) {
' +
    "    enum = execMapper('Array', enum, { ...options, id: 'Object' });
" +
    '    obj.enum = enum;
' +
    '  }
' +
    '  // Map default TestModel:eebzbch:default 
' +
    '  let default = input.default;
' +
    '  if (default !== undefined) {
' +
    "    default = execMapper('String', default, options);
" +
    '    obj.default = default;
' +
    '  }
' +
    '  // Map name TestModel:eebzbch:default 
' +
    '  let name = input.name;
' +
    '  if (name !== undefined) {
' +
    "    name = execMapper('String', name, options);
" +
    '    obj.name = name;
' +
    '  }
' +
    '  // Map description TestModel:eebzbch:default 
' +
    '  let description = input.description;
' +
    '  if (description !== undefined) {
' +
    "    description = execMapper('String', description, options);
" +
    '    obj.description = description;
' +
    '  }
' +
    '  return obj;
' +
    '}
' +
    '    at JsonSerializer.compile (/Users/vincent/workspace/test7.36/node_modules/@tsed/json-mapper/src/domain/JsonMapperCompiler.ts:168:15)
' +
    '    at JsonSerializer.map (/Users/vincent/workspace/test7.36/node_modules/@tsed/json-mapper/src/domain/JsonSerializer.ts:60:25)
' +
    '    at serialize (/Users/vincent/workspace/test7.36/node_modules/@tsed/json-mapper/src/utils/serialize.ts:11:21)
' +
    '    at PlatformResponseFilter.serialize (/Users/vincent/workspace/test7.36/node_modules/@tsed/platform-response-filter/src/services/PlatformResponseFilter.ts:92:25)
' +
    '    at PlatformHandler.flush (/Users/vincent/workspace/test7.36/node_modules/@tsed/common/src/services/PlatformHandler.ts:98:44)
' +
    '    at /Users/vincent/workspace/test7.36/node_modules/@tsed/common/src/services/PlatformHandler.ts:184:51
' +
    '    at catchAsyncError (/Users/vincent/workspace/test7.36/node_modules/@tsed/core/src/utils/catchError.ts:11:11)
' +
    '    at PlatformHandler.onFinish (/Users/vincent/workspace/test7.36/node_modules/@tsed/common/src/services/PlatformHandler.ts:184:39)
' +
    '    at /Users/vincent/workspace/test7.36/node_modules/@tsed/platform-express/src/components/PlatformExpress.ts:152:13
' +
    '    at runInContext (/Users/vincent/workspace/test7.36/node_modules/@tsed/di/src/node/utils/asyncHookContext.ts:24:12)'
}
```


### Repository URL example

https://github.com/Bugslive/tsed-7.36.git

### OS

macOS

### Node version

Node v16.20.2

### Library version

v7.36.x

### Additional context

_No response_

--- a/packages/specs/json-mapper/src/domain/JsonDeserializer.spec.ts
+++ b/packages/specs/json-mapper/src/domain/JsonDeserializer.spec.ts
@@ -1,10 +1,8 @@
 import faker from "@faker-js/faker";
 import {QueryParams} from "@tsed/common";
-import {cleanObject, getValue} from "@tsed/core";
 import {
   AdditionalProperties,
   CollectionOf,
-  Default,
   DiscriminatorKey,
   DiscriminatorValue,
   Email,
@@ -14,7 +12,6 @@ import {
   Groups,
   Ignore,
   In,
-  Integer,
   JsonEntityStore,
   JsonHookContext,
   JsonParameterStore,
@@ -26,14 +23,12 @@ import {
   Property,
   Required
 } from "@tsed/schema";
-import {snakeCase} from "change-case";
 import {Post} from "../../test/helpers/Post";
 import {User} from "../../test/helpers/User";
 import "../components/DateMapper";
 import "../components/PrimitiveMapper";
 import "../components/SymbolMapper";
 import {OnDeserialize} from "../decorators/onDeserialize";
-import {OnSerialize} from "../decorators/onSerialize";
 import {JsonDeserializer} from "./JsonDeserializer";
 
 const deserializer = new JsonDeserializer();
@@ -246,6 +241,54 @@ describe("deserialize()", () => {
       expect(result).toBeInstanceOf(Model);
       expect(result).toEqual({
         prop: [1, 2, 3, 5]
+      });
+    });
+    it("should transform object to class (reserved keyword)", () => {
+      class Role {
+        @Property()
+        enum: string;
+
+        constructor({enum: en}: any = {}) {
+          this.enum = en;
+        }
+      }
+
+      class Model {
+        @Property()
+        id: string;
+
+        @Ignore((ignored, ctx: JsonHookContext) => ctx.api)
+        password: string;
+
+        @OnDeserialize((value) => String(value) + "test")
+        @Name("mapped_prop")
+        mappedProp: string;
+
+        @CollectionOf(Role)
+        roles: Map<string, Role> = new Map();
+      }
+
+      const result = deserialize(
+        {
+          id: "id",
+          password: "string",
+          mapped_prop: "mappedProp",
+          roles: {
+            role1: {
+              enum: "role"
+            }
+          },
+          add: true
+        },
+        {type: Model, additionalProperties: false}
+      );
+
+      expect(result).toBeInstanceOf(Model);
+      expect(result).toEqual({
+        id: "id",
+        mappedProp: "mappedProptest",
+        password: "string",
+        roles: new Map().set("role1", new Role({enum: "role"}))
       });
     });
     it("should transform object to class (additionalProperties = false)", () => {

--- a/packages/specs/json-mapper/src/domain/JsonDeserializer.ts
+++ b/packages/specs/json-mapper/src/domain/JsonDeserializer.ts
@@ -17,6 +17,10 @@ function isDeserializable(obj: any, options: JsonDeserializerOptions) {
   return !(isEmpty(options.type) || (options.type === Object && !options.collectionType));
 }
 
+function varKey(k: string) {
+  return `__${k}`;
+}
+
 function mapParamStoreOptions(store: JsonParameterStore, options: JsonDeserializerOptions) {
   return {
     ...options,
@@ -220,17 +224,17 @@ export class JsonDeserializer extends JsonMapperCompiler<JsonDeserializerOptions
       getter = `alterValue('${schemaId}', input[${pick}], ${opts})`;
     }
 
-    const ifWriter = writer.set(`let ${key}`, getter).if(`${key} !== undefined`);
+    const ifWriter = writer.set(`let ${varKey(key)}`, getter).if(`${varKey(key)} !== undefined`);
 
     const fill = this.getPropertyFiller(propertyStore, id, groups, formatOpts);
 
     if (hasDeserializer) {
-      fill(ifWriter.if(`${key} === input.${key}`));
+      fill(ifWriter.if(`${varKey(key)} === input.${key}`));
     } else {
       fill(ifWriter);
     }
 
-    ifWriter.set(`obj.${key}`, key);
+    ifWriter.set(`obj.${key}`, varKey(key));
 
     if (groups && groups.includes("partial")) {
       ifWriter.else().add(`delete obj.${key}`);
@@ -249,23 +253,25 @@ export class JsonDeserializer extends JsonMapperCompiler<JsonDeserializerOptions
       const index = getGenericIndex(propertyStore);
       const opts = Writer.options(formatOpts, `type: generics[${index}]`);
 
-      return (writer: Writer) => writer.set(key, `compileAndMap(${key}, ${opts})`);
+      return (writer: Writer) => writer.set(varKey(key), `compileAndMap(${varKey(key)}, ${opts})`);
     }
 
     const type = propertyStore.itemSchema.hasDiscriminator ? propertyStore.itemSchema.discriminator().base : propertyStore.getBestType();
     const nestedMapper = this.compile(type, groups);
 
     if (propertyStore.isCollection) {
-      return (writer: Writer) => writer.callMapper(nameOf(propertyStore.collectionType), key, `id: '${nestedMapper.id}'`, formatOpts);
+      return (writer: Writer) =>
+        writer.callMapper(nameOf(propertyStore.collectionType), varKey(key), `id: '${nestedMapper.id}'`, formatOpts);
     }
 
     if (generics?.length) {
       this.schemes[schemaId] = propertyStore.schema;
 
-      return (writer: Writer) => writer.callMapper(nestedMapper.id, key, formatOpts, `generics: schemes['${schemaId}'].nestedGenerics`);
+      return (writer: Writer) =>
+        writer.callMapper(nestedMapper.id, varKey(key), formatOpts, `generics: schemes['${schemaId}'].nestedGenerics`);
     }
 
-    return (writer: Writer) => writer.callMapper(nestedMapper.id, key, formatOpts);
+    return (writer: Writer) => writer.callMapper(nestedMapper.id, varKey(key), formatOpts);
   }
 
   private mapOptions({groups = false, useAlias = true, types, ...options}: JsonDeserializerOptions<any, any>): JsonDeserializerOptions {

--- a/packages/specs/json-mapper/src/domain/JsonSerializer.spec.ts
+++ b/packages/specs/json-mapper/src/domain/JsonSerializer.spec.ts
@@ -788,6 +788,77 @@ describe("JsonSerializer", () => {
         }
       ]);
     });
+    it("should serialize model (protected keyword)", () => {
+      class Role {
+        @Property()
+        label: string;
+
+        constructor({label}: any = {}) {
+          this.label = label;
+        }
+      }
+
+      class Model {
+        @Property()
+        id: string;
+
+        @Ignore((ignored, ctx: JsonHookContext) => ctx.api)
+        password: string;
+
+        @OnSerialize((value) => String(value) + "test")
+        @Name("mapped_prop")
+        mappedProp: string;
+
+        @Property()
+        enum: string;
+
+        @CollectionOf(Role)
+        roles: Map<string, Role> = new Map();
+      }
+
+      const model = new Model();
+      // @ts-ignore
+      model.$isMongooseModelPrototype = true;
+      // @ts-ignore
+      model["toJSON"] = (options: any) => {
+        const result = deserialize(
+          {
+            id: "id",
+            password: "hellopassword",
+            mappedProp: "hello",
+            enum: "test"
+          },
+          {useAlias: false, type: Model}
+        );
+
+        return serialize(result, options);
+      };
+
+      expect(serialize(model, {type: Model})).toEqual({
+        id: "id",
+        enum: "test",
+        mapped_prop: "hellotest",
+        password: "hellopassword",
+        roles: {}
+      });
+
+      expect(serialize(model, {api: true, useAlias: false})).toEqual({
+        id: "id",
+        enum: "test",
+        mappedProp: "hellotest",
+        roles: {}
+      });
+
+      expect(serialize([model], {type: Model})).toEqual([
+        {
+          id: "id",
+          enum: "test",
+          mapped_prop: "hellotest",
+          password: "hellopassword",
+          roles: {}
+        }
+      ]);
+    });
     it("should serialize model Array", () => {
       class Role {
         @Property()

--- a/packages/specs/json-mapper/src/domain/JsonSerializer.ts
+++ b/packages/specs/json-mapper/src/domain/JsonSerializer.ts
@@ -36,6 +36,10 @@ function getBestType(type: Type<any>, obj: any) {
   return type || Object;
 }
 
+function varKey(k: string) {
+  return `__${k}`;
+}
+
 export class JsonSerializer extends JsonMapperCompiler<JsonSerializerOptions> {
   constructor() {
     super();
@@ -156,20 +160,20 @@ export class JsonSerializer extends JsonMapperCompiler<JsonSerializerOptions> {
       getter = `alterValue('${schemaId}', input.${key}, ${opts})`;
     }
 
-    writer = writer.set(`let ${key}`, getter).if(`${key} !== undefined`);
+    writer = writer.set(`let ${varKey(key)}`, getter).if(`${varKey(key)} !== undefined`);
 
     const fill = this.getPropertyFiller(propertyStore, key, groups, formatOpts);
 
     if (hasSerializer) {
-      fill(writer.if(`${key} === input.${key}`));
+      fill(writer.if(`${varKey(key)} === input.${key}`));
     } else {
       fill(writer);
     }
 
     if (aliasKey !== key) {
-      writer.set(`obj[options.useAlias ? '${aliasKey}' : '${key}']`, key);
+      writer.set(`obj[options.useAlias ? '${aliasKey}' : '${key}']`, varKey(key));
     } else {
-      writer.set(`obj.${key}`, key);
+      writer.set(`obj.${key}`, varKey(key));
     }
 
     return writer.root();
@@ -181,13 +185,14 @@ export class JsonSerializer extends JsonMapperCompiler<JsonSerializerOptions> {
 
       const nestedMapper = this.compile(type, groups);
 
-      return (writer: Writer) => writer.callMapper(nameOf(propertyStore.collectionType), key, `id: '${nestedMapper.id}'`, formatOpts);
+      return (writer: Writer) =>
+        writer.callMapper(nameOf(propertyStore.collectionType), varKey(key), `id: '${nestedMapper.id}'`, formatOpts);
     }
 
     const type = propertyStore.getBestType();
     const nestedMapper = this.compile(type, groups);
 
-    return (writer: Writer) => writer.callMapper(nestedMapper.id, key, formatOpts);
+    return (writer: Writer) => writer.callMapper(nestedMapper.id, varKey(key), formatOpts);
   }
 
   private mapPrecondition(id: string) {

